### ISSUE TITLE
False positive when `dataclass()` decorator is stored in variable and then applied

### ISSUE THREAD

**Issue by denballakh** [microsoft/pyright#8230]
**Describe the bug**
```py
from dataclasses import dataclass

# deco = dataclass # if i use this - everything is fine
deco = dataclass()  # this causes error on `x = X('foo')` line
# deco = dataclass(repr=False)  # with any arguments it causes errors as well

# at runtime all variants work without any errors

@deco
class X:
    name: str

x = X('foo')  # error: Expected no arguments to "X" constructor
print(x.name)
```

**VS Code extension or command-line**
```sh
>pyright --version
pyright 1.1.369
```

**Comment by erictraut** [comment]
Pyright's behavior here is intended, so this isn't a bug.

The `dataclass` decorator requires significant special-casing in static analysis tools, so it must be applied directly to the class that you are decorating. When you use an intermediate variable like `deco` in your example, information is lost. This isn't specific to pyright; it's true for other static type checkers like mypy.

If you look at the type of `deco`, you'll see that it's simply a function that accepts `type[T]` and returns `type[T]` with no additional changes.

If you want your code to work with static type checking, you'll need to apply the `dataclass()` decorator directly to the class.

```python
@dataclass()
class X:
    name: str
```
